### PR TITLE
Test forked joins

### PR DESCRIPTION
**Pull request by gmaclennan**
@noffle can you review that my test makes sense? Should we merge this test into master?

### CODE REVIEW

File: test/forked_joins.js [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+var osmdb = require('../')
+var versions = {}
+var way0Key
+var osm = osmdb({
+  log: hyperlog(memdb(), { valueEncoding: 'json' }),
+  db: memdb(),
+  store: fdstore(4096, storefile)
+})
+
+test('setup db', function (t) {
```
Comment (noffle) [inline test/forked_joins.js]: -1 to this style of testing: requires shared global mutable state, so tests are highly dependent on order and presence. I prefer contained, pure tests like [this](https://github.com/digidem/osm-p2p-db/blob/master/test/fork.js), where helper logic is broken out as (non-global statal) functions outside the test.

File: test/forked_joins.js [hunk 2]
```diff
@@ -0,0 +1,168 @@
+var test = require('tape')
```
Comment (noffle) [inline test/forked_joins.js]: This is a great test to have. Nice stuff Gregor. :)

File: test/forked_joins.js [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+      t.error(err)
+      t.equal(refs.length, 2, 'A referenced by way0 and way1')
+    })
+    osm.refs.list('B', function (err, refs) {
+      t.error(err)
+      t.equal(refs.length, 0, 'B no longer referenced')
+    })
+    osm.refs.list('C', function (err, refs) {
+      t.error(err)
+      t.equal(refs.length, 1, 'C referenced by way0')
```
Comment (noffle) [inline test/forked_joins.js]: How do you know this is `way0`? Can we explicitly check somehow?

File: test/forked_joins.js [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+    })
+    osm.refs.list('D', function (err, refs) {
+      t.error(err)
+      t.equal(refs.length, 1, 'D referenced by way1')
```
Comment (noffle) [inline test/forked_joins.js]: Same here!

**Comment by gmaclennan** [comment]
fixed the test, however, it does not pass, and I think it should.

File: test/forked_joins.js (hunk 1 above)
Comment (gmaclennan) [inline test/forked_joins.js]: I kind of agree... however, the contained test you wrote is the same as just putting everything in the same file. Having each step as a test is just a cheap way to get async functions to run in order. The advantage is that it enforces order to be top to bottom, whereas writing how you did could end up with the async flow being in different directions if you are not careful. I tend to write tests like this where each `test()` is a step in an async flow, with a setup and teardown step at start and end.

Comment (noffle) [inline test/forked_joins.js]: The file I linked to isn't ideal either -- something like [run-series](https://github.com/feross/run-series) would give the best of both worlds.

File: test/forked_joins.js [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+    { type: 'put', key: 'F', value: { type: 'way', refs: [ 'A', 'B' ] } }
+  ]
+  osm.batch(batch, function (err, nodes) {
+    t.error(err)
+    nodes.forEach(function (node) {
+      versions[node.value.k] = node.key
+    })
+    ways[0] = versions.F
+    console.log('way0 version:', ways[0])
+    ready()
```
Comment (noffle) [inline test/forked_joins.js]: The indexes ought to be waited on before reading each time, right? So maybe `osm.ready(ready)`?

**Comment by noffle** [comment]
Nice! I'll ping you here once I fix the refs bug.

**Comment by gmaclennan** [comment]
We were waiting on a fix to the "refs bug" here, current status?

**Comment by noffle** [comment]
Hey, I haven't forgotten about this!

I loaded up this PR and took a look. I fixed up a few things:

1. Rebased it onto `master`
2. Changed the test to use `create_db.js` instead of the usual setup boilerplate
3. Fixed a use of `osm.del(key, {keys: ...})` to be the correct `osm.del(key, {links: ...})`
4. Fixed osm-p2p-db to handle refs correctly when there's a deletion of a deletion document (hadn't thought of this one--nice!)

Check out especially my last commit of the bunch, (5): normally you'd expect that if you delete a way, `osm.refs.list(key)` wouldn't return that way, right? Well, due to https://github.com/digidem/osm-p2p-db/pull/44, we now surface deletions, which means `osm.refs.list(key)` will return the deleted way's tombstone document. The changes in (5) tweak your tests to expect this.

The big Q: is this behaviour *desirable*? I can't remember if we discussed this before, but since I wrote the code that does it, I wonder if we were in agreement about it.

If you're feelings are a resounding "yes!", then I say we're good to merge. :tada:

**Comment by gmaclennan** [comment]
I have been thinking about this behaviour more, specifically refs on deleted ways/relations. The OSM API does not return refs on a deleted node, and from a user perspective it is confusing which nodes are returned on a deleted way from a fork (like with `way4` in the test). However, the behaviour we do want is all heads, including deleted heads, being returned from queries. 

From a user perspective, the behaviour I think we want is that a bbox query returns deleted ways/relations if there are >1 heads (a fork) and one of them is not deleted. However, an alternative might be to return the non-deleted way(s)/relation(s) which some kind of attribute that indicates that there is a (deleted) fork, and the client app can then manage lookup of the deleted way/relation.

This is tied to how we index deleted ways. I think current behavour will kind of work how we want it to, since I think we're maintaining refs on deleted ways/relations. Let's discuss this more in person this week.

**Pull request merged**